denops/@ddc-sources/path.ts

```diff
@@ -19,33 +19,68 @@ type Params = {
   cmd: string[];
   absolute: boolean;
   dirSeparator: DirSeparator | "";
+  escapeChars: string;
 };
+type WordFilter = (word: string) => string;
 
-function isValidUnixPath(path: string): boolean {
-  let escape = false;
-  for (let i = 0; i < path.length; i++) {
-    if (!escape && path[i].match(/\s/)) {
+function isValidUnixPath(path: string, escapeChars: string): boolean {
+  for (let i = 0; i < path.length; ++i) {
+    if (
+      path[i] === "\\" && (i + 1) < path.length &&
+      escapeChars.includes(path[i + 1])
+    ) {
+      ++i;
+    } else if (escapeChars.includes(path[i])) {
       return false;
     }
-    if (path[i] === "\\") {
-      escape = !escape;
-    } else {
-      escape = false;
+  }
+  return true;
+}
+
+function isValidWindowsPath(path: string, escapeChars: string): boolean {
+  // allow both '/' and '\'
+  const invalidChars = ':*?"<>|' + escapeChars;
+  let i = 0;
+  if (/^[a-zA-Z]:/.test(path)) {
+    i = 2;
+  }
+  for (; i < path.length; ++i) {
+    if (
+      path[i] === "\\" && (i + 1) < path.length &&
+      escapeChars.includes(path[i + 1])
+    ) {
+      ++i;
+    } else if (invalidChars.includes(path[i])) {
+      return false;
     }
   }
   return true;
 }
 
+function getWordFilter(escapeChars: string): WordFilter {
+  if (escapeChars.length === 0) {
+    return (word) => word;
+  }
+  const escapeRegex = new RegExp(
+    `[${escapeChars.replaceAll(/[-\\\]]/g, "\\$&")}]`,
+    "g",
+  );
+  return (word) => word.replaceAll(escapeRegex, "\\$&");
+}
+
 export class Source extends BaseSource<Params, UserData> {
-  override getCompletePosition(
-    arg: GetCompletePositionArguments<Params>,
+  override async getCompletePosition(
+    { context, denops, sourceParams }: GetCompletePositionArguments<Params>,
   ): Promise<number> {
-    for (let i = 0; i < arg.context.input.length; i++) {
-      if (isValidUnixPath(arg.context.input.slice(i))) {
-        return Promise.resolve(i);
+    const isValidPath = Deno.build.os === "windows"
+      ? isValidWindowsPath
+      : isValidUnixPath;
+    for (let i = 0; i < context.input.length; i++) {
+      if (isValidPath(context.input.slice(i), sourceParams.escapeChars)) {
+        return i;
       }
     }
-    return Promise.resolve(arg.context.input.length);
+    return context.input.length;
```

Comment:
`context.input.length` is `number` value. However, we declare that the function returns a value as a `Promise<number>`. Consider wrapping `Promise.resolve`

Reply:
`getCompletePosition()` is changed to async function, so `return <number>` is correct.
```diff
-  override getCompletePosition(
+  override async getCompletePosition(
```

Reply:
I found another bug in `getCompletePosition()`, so fix it and correct this.

Reply:
Noted. Thanks

Reply:
Thanks.